### PR TITLE
Unzip/Deflate content on error status for Default Client

### PR DESCRIPTION
**Pull request by gdufrene**
### Summary by CodeRabbit

- Refactor: Simplified the handling of gzip and deflate encoded responses in the core application. This change enhances the software's reliability by ensuring consistent handling of encoded responses.
- Test: Added new tests to verify the correct handling of gzip and deflate encoded error responses. These tests improve the robustness of the software by ensuring that it behaves as expected even in error scenarios.
- Test: Updated various client tests to include checks for gzip and deflate encoding error handling. This ensures that the software's performance remains consistent across different client implementations.

### CODE REVIEW

**Comment by gdufrene** [comment]
This MR is about processing gzip or deflate content when error status is >= 400 on default http client.
This is related to issue #1386 

Hope it would help solving this issue from 2021 :)

File: jaxrs2/src/test/java/feign/jaxrs2/JAXRSClientTest.java [hunk 1]
```diff
@@ -152,11 +152,21 @@ public void canSupportGzip() throws Exception {
...
   }
 
+  @Override
+  public void canSupportDeflateOnError() throws Exception {
+    assumeFalse("JaxRS client do not support deflate compression", false);
+  }
+
   @Override
   public void canExceptCaseInsensitiveHeader() throws Exception {
     assumeFalse("JaxRS client do not support gzip compression", false);
```
[the remark this answers is not in the document]
Comment (gdufrene) [reply jaxrs2/src/test/java/feign/jaxrs2/JAXRSClientTest.java:172]: Part of existing code, should be fixed in a separate PR if required.

File: okhttp/src/test/java/feign/okhttp/OkHttpClientTest.java [hunk 2]
```diff
@@ -115,11 +115,21 @@ public void canSupportGzip() throws Exception {
...
     assumeFalse("OkHTTP client do not support deflate compression", false);
   }
 
+  @Override
+  public void canSupportDeflateOnError() throws Exception {
+    assumeFalse("OkHTTP client do not support gzip compression", false);
+  }
+
   @Override
   public void canExceptCaseInsensitiveHeader() throws Exception {
```
[the remark this answers is not in the document]
Comment (gdufrene) [reply okhttp/src/test/java/feign/okhttp/OkHttpClientTest.java:134]: comment above methods.

**Comment by velo** [comment]
> Hope it would help solving this issue from 2021 :)

Plenty more from well dates waiting for community fixes.

**Comment by gdufrene** [comment]
> Plenty more from well dates waiting for community fixes.

And this is probably not the most awaited one.
But still, first contribution on that project, have to start small :)

Hope I will help for others in the future.

**Comment by gdufrene** [comment]
rebased this last changes on master branch.
Let me know if anything else needs to be done to allow merging this fix.

**Review by velo — APPROVED** [review]

**Pull request merged**